### PR TITLE
Leverage defineEnv for all environment variables in Turbopack

### PR DESCRIPTION
**Pull request by timneutkens**
This ensures all `process.env` replacement comes from define-env in Next.js, instead of also applying it partially on the Rust side.

### CODE REVIEW

**Comment by timneutkens** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #57196** <a href="https://app.graphite.dev/github/pr/vercel/next.js/57196" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/57196?utm_source=stack-comment).

**Comment by ijjk** [comment]
## Failing test suites <!--17490208302-->
Commit: 1c2d6553ff66026119dbe969fc51612510c7a9ea

`pnpm test test/integration/amphtml-custom-validator/test/index.test.js` 
- AMP Custom Validator > development mode > should run in dev mode successfully

<details>
<summary>Expand output</summary>

  ● AMP Custom Validator › development mode › should run in dev mode successfully

    expect(received).not.toContain(expected) // indexOf

    Expected substring: not "error"
    Received string:        "  ⚠ _app error - [project]/packages/next/dist/compiled/gzip-size/index.js - Error resolving commonjs request
    unable to resolve module \"fs\"

     error - packages/next/dist/compiled/gzip-size/index.js:1:2833  Error resolving commonjs request

    > 1 | (()=>{var e={154:(e,r,t)=>{var n=t(781);var o=[\"write\",\"end\",\"destroy\"];var i=[\"resume\",\"pause\"];var s=[\"data\",\"close\"];var c=Array.prototype.slice;e.exports=duplex;function forEach(e,r){if(e.forEach){return e.forEach(r)}for(var t=0;t<e.length;t++){r(e[t],t)}}function duplex(e,r){var t=new n;var a=false;forEach(o,proxyWriter);forEach(i,proxyReader);forEach(s,proxyStream);r.on(\"end\",handleEnd);e.on(\"drain\",(function(){t.emit(\"drain\")}));e.on(\"error\",reemit);r.on(\"error\",reemit);t.writable=e.writable;t.readable=r.readable;return t;function proxyWriter(r){t[r]=method;function method(){return e[r].apply(e,arguments)}}function proxyReader(e){t[e]=method;function method(){t.emit(e);var n=r[e];if(n){return n.apply(r,arguments)}r.emit(e)}}function proxyStream(e){r.on(e,reemit);function reemit(){var r=c.call(arguments);r.unshift(e);t.emit.apply(t,r)}}function handleEnd(){if(a){return}a=true;var e=c.call(arguments);e.unshift(\"end\");t.emit.apply(t,e)}function reemit(e){t.emit(\"error\",e)}}},349:(e,r,t)=>{\"use strict\";const n=t(147);const o=t(781);const i=t(796);const s=t(154);const c=t(530);const getOptions=e=>Object.assign({level:9},e);e.exports=(e,r)=>{if(!e){return Promise.resolve(0)}return c(i.gzip)(e,getOptions(r)).then((e=>e.length)).catch((e=>0))};e.exports.sync=(e,r)=>i.gzipSync(e,getOptions(r)).length;e.exports.stream=e=>{const r=new o.PassThrough;const t=new o.PassThrough;const n=s(r,t);let c=0;const a=i.createGzip(getOptions(e)).on(\"data\",(e=>{c+=e.length})).on(\"error\",(()=>{n.gzipSize=0})).on(\"end\",(()=>{n.gzipSize=c;n.emit(\"gzip-size\",c);t.end()}));r.pipe(a);r.pipe(t,{end:false});return n};e.exports.file=(r,t)=>new Promise(((o,i)=>{const s=n.createReadStream(r);s.on(\"error\",i);const c=s.pipe(e.exports.stream(t));c.on(\"error\",i);c.on(\"gzip-size\",o)}));e.exports.fileSync=(r,t)=>e.exports.sync(n.readFileSync(r),t)},530:e=>{\"use strict\";const processFn=(e,r)=>function(...t){const n=r.promiseModule;return new n(((n,o)=>{if(r.multiArgs){t.push(((...e)=>{if(r.errorFirst){if(e[0]){o(e)}else{e.shift();n(e)}}else{n(e)}}))}else if(r.errorFirst){t.push(((e,r)=>{if(e){o(e)}else{n(r)}}))}else{t.push(n)}e.apply(this,t)}))};e.exports=(e,r)=>{r=Object.assign({exclude:[/.+(Sync|Stream)$/],errorFirst:true,promiseModule:Promise},r);const t=typeof e;if(!(e!==null&&(t===\"object\"||t===\"function\"))){throw new TypeError(`Expected \\`input\\` to be a \\`Function\\` or \\`Object\\`, got \\`${e===null?\"null\":t}\\``)}const filter=e=>{const match=r=>typeof r===\"string\"?e===r:r.test(e);return r.include?r.include.some(match):!r.exclude.some(match)};let n;if(t===\"function\"){n=function(...t){return r.excludeMain?e(...t):processFn(e,r).apply(this,t)}}else{n=Object.create(Object.getPrototypeOf(e))}for(const t in e){const o=e[t];n[t]=typeof o===\"function\"&&filter(t)?processFn(o,r):o}return n}},147:e=>{\"use strict\";e.exports=require(\"fs\")},781:e=>{\"use strict\";e.exports=require(\"stream\")},796:e=>{\"use strict\";e.exports=require(\"zlib\")}};var r={};function __nccwpck_require__(t){var n=r[t];if(n!==undefined){return n.exports}var o=r[t]={exports:{}};var i=true;try{e[t](o,o.exports,__nccwpck_require__);i=false}finally{if(i)delete r[t]}return o.exports}if(typeof __nccwpck_require__!==\"undefined\")__nccwpck_require__.ab=__dirname+\"/\";var t=__nccwpck_require__(349);module.exports=t})();
        |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  ^^^^^^^^^^^^^
    unable to resolve module \"fs\"
    It was not possible to find the requested file.
        Parsed request as written in source code: module \"fs\"
        Path where resolving has started: [project]/packages/next/dist/compiled/gzip-size/index.js
        Type of request: commonjs request
        Import map: No import map entry
        


      ⚠ / error - [project]/packages/next/dist/compiled/gzip-size/index.js - Error resolving commonjs request
    unable to resolve module \"fs\"

     error - packages/next/dist/compiled/gzip-size/index.js:1:2833  Error resolving commonjs request

    > 1 | (()=>{var e={154:(e,r,t)=>{var n=t(781);var o=[\"write\",\"end\",\"destroy\"];var i=[\"resume\",\"pause\"];var s=[\"data\",\"close\"];var c=Array.prototype.slice;e.exports=duplex;function forEach(e,r){if(e.forEach){return e.forEach(r)}for(var t=0;t<e.length;t++){r(e[t],t)}}function duplex(e,r){var t=new n;var a=false;forEach(o,proxyWriter);forEach(i,proxyReader);forEach(s,proxyStream);r.on(\"end\",handleEnd);e.on(\"drain\",(function(){t.emit(\"drain\")}));e.on(\"error\",reemit);r.on(\"error\",reemit);t.writable=e.writable;t.readable=r.readable;return t;function proxyWriter(r){t[r]=method;function method(){return e[r].apply(e,arguments)}}function proxyReader(e){t[e]=method;function method(){t.emit(e);var n=r[e];if(n){return n.apply(r,arguments)}r.emit(e)}}function proxyStream(e){r.on(e,reemit);function reemit(){var r=c.call(arguments);r.unshift(e);t.emit.apply(t,r)}}function handleEnd(){if(a){return}a=true;var e=c.call(arguments);e.unshift(\"end\");t.emit.apply(t,e)}function reemit(e){t.emit(\"error\",e)}}},349:(e,r,t)=>{\"use strict\";const n=t(147);const o=t(781);const i=t(796);const s=t(154);const c=t(530);const getOptions=e=>Object.assign({level:9},e);e.exports=(e,r)=>{if(!e){return Promise.resolve(0)}return c(i.gzip)(e,getOptions(r)).then((e=>e.length)).catch((e=>0))};e.exports.sync=(e,r)=>i.gzipSync(e,getOptions(r)).length;e.exports.stream=e=>{const r=new o.PassThrough;const t=new o.PassThrough;const n=s(r,t);let c=0;const a=i.createGzip(getOptions(e)).on(\"data\",(e=>{c+=e.length})).on(\"error\",(()=>{n.gzipSize=0})).on(\"end\",(()=>{n.gzipSize=c;n.emit(\"gzip-size\",c);t.end()}));r.pipe(a);r.pipe(t,{end:false});return n};e.exports.file=(r,t)=>new Promise(((o,i)=>{const s=n.createReadStream(r);s.on(\"error\",i);const c=s.pipe(e.exports.stream(t));c.on(\"error\",i);c.on(\"gzip-size\",o)}));e.exports.fileSync=(r,t)=>e.exports.sync(n.readFileSync(r),t)},530:e=>{\"use strict\";const processFn=(e,r)=>function(...t){const n=r.promiseModule;return new n(((n,o)=>{if(r.multiArgs){t.push(((...e)=>{if(r.errorFirst){if(e[0]){o(e)}else{e.shift();n(e)}}else{n(e)}}))}else if(r.errorFirst){t.push(((e,r)=>{if(e){o(e)}else{n(r)}}))}else{t.push(n)}e.apply(this,t)}))};e.exports=(e,r)=>{r=Object.assign({exclude:[/.+(Sync|Stream)$/],errorFirst:true,promiseModule:Promise},r);const t=typeof e;if(!(e!==null&&(t===\"object\"||t===\"function\"))){throw new TypeError(`Expected \\`input\\` to be a \\`Function\\` or \\`Object\\`, got \\`${e===null?\"null\":t}\\``)}const filter=e=>{const match=r=>typeof r===\"string\"?e===r:r.test(e);return r.include?r.include.some(match):!r.exclude.some(match)};let n;if(t===\"function\"){n=function(...t){return r.excludeMain?e(...t):processFn(e,r).apply(this,t)}}else{n=Object.create(Object.getPrototypeOf(e))}for(const t in e){const o=e[t];n[t]=typeof o===\"function\"&&filter(t)?processFn(o,r):o}return n}},147:e=>{\"use strict\";e.exports=require(\"fs\")},781:e=>{\"use strict\";e.exports=require(\"stream\")},796:e=>{\"use strict\";e.exports=require(\"zlib\")}};var r={};function __nccwpck_require__(t){var n=r[t];if(n!==undefined){return n.exports}var o=r[t]={exports:{}};var i=true;try{e[t](o,o.exports,__nccwpck_require__);i=false}finally{if(i)delete r[t]}return o.exports}if(typeof __nccwpck_require__!==\"undefined\")__nccwpck_require__.ab=__dirname+\"/\";var t=__nccwpck_require__(349);module.exports=t})();
        |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  ^^^^^^^^^^^^^
    unable to resolve module \"fs\"
    It was not possible to find the requested file.
        Parsed request as written in source code: module \"fs\"
        Path where resolving has started: [project]/packages/next/dist/compiled/gzip-size/index.js
        Type of request: commonjs request
        Import map: No import map entry
        


    "

      45 |       await killApp(app)
      46 |
    > 47 |       expect(stderr).not.toContain('error')
         |                          ^
      48 |       expect(html).toContain('Hello from AMP')
      49 |     })
      50 |   })

      at Object.toContain (integration/amphtml-custom-validator/test/index.test.js:47:26)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 10-21-Leverage_defineEnv_for_all_environment_variables_in_Turbopack | Change |
| - | - | - | - |
| buildDuration | 10.3s | 10.3s | N/A |
| buildDurationCached | 6.1s | 6s | N/A |
| nodeModulesSize | 174 MB | 174 MB | ⚠️ +2.2 kB |
| nextStartRea..uration (ms) | 518ms | 512ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 10-21-Leverage_defineEnv_for_all_environment_variables_in_Turbopack | Change |
| - | - | - | - |
| 199-HASH.js gzip | 27.6 kB | 27.6 kB | ✓ |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| 99.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.3 kB | 45.3 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 252 B | N/A |
| main-HASH.js gzip | 32.9 kB | 32.9 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 126 kB | 126 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 10-21-Leverage_defineEnv_for_all_environment_variables_in_Turbopack | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 10-21-Leverage_defineEnv_for_all_environment_variables_in_Turbopack | Change |
| - | - | - | - |
| _app-HASH.js gzip | 206 B | 205 B | N/A |
| _error-HASH.js gzip | 182 B | 180 B | N/A |
| amp-HASH.js gzip | 506 B | 505 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | N/A |
| edge-ssr-HASH.js gzip | 260 B | 259 B | N/A |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.36 kB | 4.36 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.63 kB | N/A |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 384 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.08 kB | 1.08 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 10-21-Leverage_defineEnv_for_all_environment_variables_in_Turbopack | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 10-21-Leverage_defineEnv_for_all_environment_variables_in_Turbopack | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 541 B | 542 B | N/A |
| withRouter.html gzip | 523 B | 524 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 10-21-Leverage_defineEnv_for_all_environment_variables_in_Turbopack | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.8 kB | 93.8 kB | N/A |
| page.js gzip | 155 kB | 155 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 10-21-Leverage_defineEnv_for_all_environment_variables_in_Turbopack | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 623 B | 622 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 22.9 kB | 22.9 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

</details>
Commit: 1c2d6553ff66026119dbe969fc51612510c7a9ea

**Comment by sokra** [comment]
Still failing

File: packages/next-swc/crates/next-core/src/next_client/context.rs [hunk 1]
```diff
@@ -64,40 +63,34 @@ use crate::{
...
+                    Ok(serde_json::Value::Bool(v)) => CompileTimeDefineValue::Bool(v),
+                    Ok(serde_json::Value::String(v)) => CompileTimeDefineValue::String(v),
+                    _ => CompileTimeDefineValue::JSON(v.clone()),
+                }
+            });
     }
 
-    defines
+    CompileTimeDefines(defines)
 }
```
Comment (sokra) [inline packages/next-swc/crates/next-core/src/next_client/context.rs:83]: This should be moved into a shared files

**Pull request merged**